sdks/typescript/src/apache_beam/worker/logging.ts

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AsyncLocalStorage, AsyncResource } from "node:async_hooks";
+
+export const loggingLocalStorage = new AsyncLocalStorage();
+
+import * as grpc from "@grpc/grpc-js";
+import { startCapture } from "capture-console";
+import { Queue } from "queue-typescript";
+
+import { LogEntry, LogEntry_Severity_Enum } from "../proto/beam_fn_api";
+import { BeamFnLoggingClient } from "../proto/beam_fn_api.grpc-client";
+
+export interface LoggingStageInfo {
+  instructionId?: string;
+  transformId?: string;
+}
+
+export class LoggingStageInfoHolder {
+  loggingStageInfo: LoggingStageInfo;
+}
+
+export function createLoggingChannel(workerId: string, endpoint: string) {
+  const logQueue = new Queue<LogEntry>();
+  function toEntry(line: string, severity: number): LogEntry {
+    const now_ms = Date.now();
+    const seconds = Math.trunc(now_ms / 1000);
+    const stageInfo = loggingLocalStorage.getStore() as LoggingStageInfo;
+    return LogEntry.create({
+      // TODO: Try to capture, or at least infer, the log level.
```

Comment:
Consider filing a P2/3 issue to link to here

Reply:
Actually, this TODO (copied from the move) is addressed now. Removed.